walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,10 +15,157 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+import java.lang.Boolean;
+import java.util.HashSet;
+import java.util.HashMap;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-    //idea use recursion over all attendees and see if you can find a start time that works for everyone
+
+  public Collection<TimeRange> calendarQuery(Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    ArrayList<Event> listEvents = new ArrayList<>(events);
+    HashSet<String> mandatoryAttendees = new HashSet<>(request.getAttendees());
+    ArrayList<Event> sortedListEvents = mySort(listEvents, mandatoryAttendees);
+
+    int rangeStart = 0;
+    TimeRange previousSlot = null;
+    for (Event event: sortedListEvents) {
+      TimeRange eventTimeRange = event.getWhen();
+      // Checks for overlapping events
+      if (previousSlot == null || !eventTimeRange.overlaps(previousSlot)) {
+        availableTimes.add(TimeRange.fromStartEnd(rangeStart, eventTimeRange.start(), false));
+        rangeStart = eventTimeRange.end();
+        previousSlot = eventTimeRange;
+      } else if(previousSlot != null && previousSlot.contains(eventTimeRange)) {
+        // Checks for nested events
+        rangeStart = previousSlot.end();
+      } else {
+        rangeStart = eventTimeRange.end();
+        previousSlot = eventTimeRange;
```

Comment:
looks like lines 46-47 are the same as 40-41. You should be able to condense these.

Reply:
done.